### DISCUSSION TITLE
Animated Apollonius diagram

### DISCUSSION THREAD

**Discussion by stla** [CGAL/cgal#7581] [Show and tell]
Hello,

Could you explain why more edges appear as the radii decrease?

![agraph03](https://github.com/CGAL/cgal/assets/4466543/3bdbb272-6250-49fa-baac-15b89f91db3c)

**Comment by sloriot** [comment]
The diagram should be a partition of the plane, so they are definitely missing edges.

I haven't looked at your code but I did wrote something for Ipe that does something similar here:
https://github.com/CGAL/cgal/blob/master/CGAL_ipelets/include/CGAL/CGAL_Ipelet_base_v7.h#L724

In theory, updating `Voronoi_from_tri` should be all you need (This is old code, please be indulgent)

**Comment by stla** [comment]
Hmm that's embarrassing.

My [code is here](https://github.com/stla/Apollonius/blob/main/src/Apollonius.cpp) but there's a part done in R and the C++ code can look complicated because originally I didn't find how to use vertices with info. But to sum up, I iterate over the finite faces and I take the dual of each face.

**Comment by lrineau** [comment]
That is because you only display finite edges. When the radii decrease, the corresponding face switch from "infinite faces" to "finite faces", and same for the edges.

**Reply by stla** [reply]
That's what I suspected, thanks. Currently trying to deal with `all_faces_begin()` instead of `finite_faces_begin()` but getting problems (see the issue I've just opened).

**Comment by stla** [comment]
So, when I iterate over *all* the faces, I get some `Line_2` as dual of the infinite faces, that's correct? I plotted these lines in blue and I'm lost, they don't help to partition the plane. Observe they are tangent to some circles, I don't know why.

![Rplot](https://github.com/CGAL/cgal/assets/4466543/224d0f91-9499-40d7-abcc-fb7fd0b9bc72)

**Comment by stla** [comment]
I'm able to get the partition without using the dual `Line_2`s. There's just a problem: I have to manually set the direction of the hyperbolic rays (the blue lines). 

Yet I don't know what are these `Line_2`s. Maybe lines of hyperbolic reflections?

![Rplot](https://github.com/CGAL/cgal/assets/4466543/90ef8d39-5440-4a2a-9de9-f05159d08c77)

**Reply by stla** [reply]
I think I get it now. An infinite face has only one finite edge. Its dual line is tangent to the two circles centered at the vertices of this edge. These two circles are on the same side of the line, and the infinite ray must remain in the opposite side.

**Comment by stla** [comment]
Yeah it works now!

![agraph03](https://github.com/CGAL/cgal/assets/4466543/76367689-0413-4ece-a26a-6d71dc216ce9)

**Comment by stla** [comment]
Golden spiral. A bit weird.

![goldenSpiral](https://github.com/CGAL/cgal/assets/4466543/cb61bc31-81eb-4bde-9e94-af74a30ea89c)

**Comment by stla** [comment]
Fermat spiral -> failure :-( Why?

![Rplot](https://github.com/CGAL/cgal/assets/4466543/630ff9ee-9908-4937-a603-b7d385e48bc4)

**Comment by stla** [comment]
Any idea why it fails? The original sites (red points) are not isolated.

![Rplot](https://github.com/CGAL/cgal/assets/4466543/66b2732a-0761-462b-bc1a-b8237a29865e)

**Comment by stla** [comment]
It works when the weights vary from 1 to 2. It fails when they vary more.

![Rplot](https://github.com/CGAL/cgal/assets/4466543/6ed27950-c1a1-4447-b709-be14e37ecdc0)

**Comment by stla** [comment]
I think some circles are contained into other circles. Can it be the reason of the failure?

![Rplot](https://github.com/CGAL/cgal/assets/4466543/6bde1674-195c-4bfe-86c8-c3aaa6525e06)

**Comment by sloriot** [comment]
Check https://doc.cgal.org/latest/Apollonius_graph_2/classCGAL_1_1Apollonius__graph__2.html#abd505d277047969aac281c3b06d3024e

**Comment by stla** [comment]
Thanks, I will check that. Let's see what happens on an animation first. The radii vary from `1` to `rmax`:

![FermatSpiral](https://github.com/CGAL/cgal/assets/4466543/f8f70e13-1623-4a47-86de-939f81c3117e)

We observe that the hyperbolic segments become highly curved. Then it is not surprising that something weird happens.

**Comment by stla** [comment]
![enclosedSquare](https://github.com/CGAL/cgal/assets/4466543/405674e0-06b3-4405-a873-c60ac6e64f75)
